druid/src/widget/tabs.rs

```diff
@@ -0,0 +1,1006 @@
+    fn new(name: impl Into<LabelText<T>>, child: impl Widget<T> + 'static) -> Self {
+        InitialTab {
+            name: SingleUse::new(name.into()),
+            child: SingleUse::new(child.boxed()),
+        }
+    }
+}
+
+enum TabsContent<TP: TabsPolicy> {
+    Building {
+        tabs: TP::Build,
+    },
+    Complete {
+        tabs: TP,
+    },
+    Running {
+        scope: WidgetPod<TP::Input, TabsScope<TP>>,
+    },
+    Swapping,
+}
+
+/// A tabs widget.
+///
+/// The tabs can be provided up front, using Tabs::new() and add_tab()/with_tab().
+///
+/// Or, the tabs can be derived from the input data by implementing TabsPolicy, and providing it to
+/// Tabs::from_policy()
+///
+/// ```
+/// use druid::widget::{Tabs, Label, WidgetExt};
+/// use druid::{Data, Lens};
+///
+/// #[derive(Data, Clone, Lens)]
+/// struct AppState{
+///     name: String
+/// }
+///
+/// let tabs = Tabs::new()
+///     .with_tab("Connection", Label::new("Connection information"))
+///     .with_tab("Proxy", Label::new("Proxy settings"))
+///     .lens(AppState::name);
+///
+///
+/// ```
+///
+pub struct Tabs<TP: TabsPolicy> {
+    axis: Axis,
+    cross: CrossAxisAlignment, // Not sure if this should have another enum. Middle means nothing here
+    transition: TabsTransition,
+    content: TabsContent<TP>,
+}
+
+impl<T: Data> Tabs<StaticTabs<T>> {
+    /// Create a new Tabs widget, using the static tabs policy.
+    /// Use with_tab or add_tab to configure the set of tabs available.
+    pub fn new() -> Self {
+        Tabs::building(Vec::new())
+    }
+}
+
+impl<T: Data> Default for Tabs<StaticTabs<T>> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<TP: TabsPolicy> Tabs<TP> {
+    fn of_content(content: TabsContent<TP>) -> Self {
+        Tabs {
+            axis: Axis::Horizontal,
+            cross: CrossAxisAlignment::Start,
+            transition: Default::default(),
+            content,
+        }
+    }
+
+    /// Create a Tabs widget using the provided policy.
+    /// This is useful for tabs derived from data.
+    pub fn for_policy(tabs: TP) -> Self {
+        Self::of_content(TabsContent::Complete { tabs })
+    }
+
+    // This could be public if there is a case for custom policies that support static tabs - ie the AddTab method.
+    // It seems very likely that the whole way we do dynamic vs static will change before that
+    // becomes an issue.
+    fn building(tabs_from_data: TP::Build) -> Self
+    where
+        TP: AddTab,
+    {
+        Self::of_content(TabsContent::Building {
+            tabs: tabs_from_data,
+        })
+    }
+
+    /// Lay out the tab bar along the provided axis.
+    pub fn with_axis(mut self, axis: Axis) -> Self {
+        self.axis = axis;
+        self
+    }
+
+    /// Put the tab bar at the corresponding end of the cross axis.
+    /// Defaults to Start. Note that Middle has the same effect as Start.
+    pub fn with_cross_axis_alignment(mut self, cross: CrossAxisAlignment) -> Self {
```

Comment:
echo my earlier comment I would prefer to call this something like `with_tab_bar_location` or whatever.

Reply:
Yep